STREAM-70 Create Snowflake schema crawler

akuma432 opened the pull request:
* added snowflake-schema-crawler templates for tables and columns
* retrieving the columns details in batch of 10 tables.
* updated filter condition for getting the schema details for different jdbc connection.

akuma432 commented:
Made some progress on this. Few things are still in progress that I will be updating tomorrow. Please share your feedback on the changes made.

akuma432 commented on src/main/java/io/phdata/streamliner/schemadefiner/util/StreamlinerUtil.java:
```diff
@@ -160,6 +161,16 @@ public static Configuration mapJdbcCatalogToConfig(Configuration ingestConfig, S
         return newConfig;
     }
 
+  private static Predicate<Schema> matchDatabaseOrSchema(Jdbc jdbc) {
```
In mysql,  database and schema both are synonym.  And in oracle or snowflake both database and schema exists. Hence this method is used just to filter schema based on jdbc connection type.

akuma432 commented on src/main/resources/snowflake-schema-crawler-tables.sql:
```diff
@@ -0,0 +1 @@
+select TABLE_SCHEMA, TABLE_NAME,TABLE_TYPE, COMMENT from information_schema.tables where UPPER(table_schema) = UPPER('{{SCHEMA_NAME}}');
```
Please review this query that retrieves tables details.

akuma432 commented on src/main/resources/snowflake-schema-crawler-columns.sql:
```diff
@@ -0,0 +1,4 @@
+select TABLE_NAME, COLUMN_NAME, DATA_TYPE, IS_NULLABLE, COMMENT, ORDINAL_POSITION,
+       DECODE(DATA_TYPE, 'NUMBER', NUMERIC_PRECISION, 'TEXT', CHARACTER_MAXIMUM_LENGTH, 'TIMESTAMP_LTZ', DATETIME_PRECISION , 'TIMESTAMP_NTZ' ,DATETIME_PRECISION, 0) as PRECISION,
+       DECODE(DATA_TYPE, 'NUMBER', NUMERIC_SCALE,  0) as SCALE
+ from information_schema.columns where table_schema = '{{SCHEMA_NAME}}' AND table_name IN ({{TABLE_LIST}}) ;
```
Please review this query that retrieves columns details

brockn replied:
Nice. I was going to correct your use of PRECISION above but it's right.

brockn commented on src/main/java/schemacrawler/crawl/StreamlinerSchemaCrawler.java:
```diff
@@ -193,6 +208,86 @@ public static StreamlinerCatalog getCatalog(String schemaName, String jdbcUrl,
+    }
+    {
+      int snowflakeTablesCount = tableList.size();
+      for (int i = 0; i < snowflakeTablesCount; i = i + 10) {
```
I was suspect of this logic with batches of say 11 or 21, but I think it's right.

https://docs.oracle.com/javase/tutorial/java/nutsandbolts/for.html

akuma432 replied:
added a parameter "batchTableCount" in config file to change this value.

brockn commented:
Excellent work, this looks really good. 

Have you tested it on Snowflake?

akuma432 commented:
> Excellent work, this looks really good.
> 
> Have you tested it on Snowflake?

Yes it is tested on Snowflake by providing only database in jdbc url.
Also I have added a parameter "batchTableCount" to change the batch table value. And added SNOWFLAKE_SCHEMAS_TEMPLATE to validate the schema provided in config.
